### PR TITLE
Use servant-client to make federated calls

### PR DESCRIPTION
**Pull request by akshaymankar**

### CODE REVIEW

File: hack/federation/grpcurlhandle.sh [hunk 1]
```diff
@@ -15,8 +16,8 @@ function getHandle() {
         AUTHORITY="-authority $SERVERNAME"
     fi
     set -x
-    grpcurl -d @ -format json $AUTHORITY $MODE -proto ../../libs/wire-api-federation/proto/router.proto "$HOST:$PORT" wire.federator.Inward/call <<EOM
-{"method": "GET", "component": "Brig", "path": "$path", "query": [{"key":"$queryK", "value":"$queryV"}]}
+    grpcurl -d @ -format json $AUTHORITY $MODE -import-path "${TOP_LEVEL}/libs/wire-api-federation/proto/" -proto router.proto "$HOST:$PORT" wire.federator.Inward/call <<EOM
+{"component": "Brig", "path": "$path"}
```
Comment (jschaul) [inline hack/federation/grpcurlhandle.sh:20]: where did the handle=alice part go?

Comment (akshaymankar) [reply hack/federation/grpcurlhandle.sh]: Added it back.

File: services/federator/src/Federator/ExternalServer.hs [hunk 2]
```diff
@@ -47,11 +49,15 @@ import Wire.API.Federation.GRPC.Types
...
-  pure $ InwardResponseHTTPResponse $ HTTPResponse statusW32 bodyBS
+  (resStatus, resBody) <- brigCall path body
+  pure $ case HTTP.statusCode resStatus of
+    200 -> InwardResponseBody $ maybe mempty LBS.toStrict resBody
+    -- TODO: There is a unit test for this, but Akshay has seen the integration
+    -- test never sees InwardResponseErr, when the error is supposed to be
+    -- returned, the integration test just sees `InwardResponseBody` with empty
+    -- body. Maybe this is a bug in mu-haskell, maybe something is wrong with
+    -- our integration test, let's verify this.
+    code -> InwardResponseErr $ "Invalid HTTP status from component: " <> Text.pack (show code) <> " " <> Text.decodeUtf8 (HTTP.statusMessage resStatus)
```
Comment (akshaymankar) [inline services/federator/src/Federator/ExternalServer.hs:60]: TODO: Add a log statement here as brig should ideally never return something that is not 200.

Comment (akshaymankar) [reply services/federator/src/Federator/ExternalServer.hs:60]: Actually brig can return 400 when parsing the request body fails.

**Comment by smatting** [comment]
Integration tests with unknown domain fail because `gRpcCall` (https://github.com/wireapp/wire-server/pull/1445/files#diff-72304808ff3781aece246f66abc167cd8b9b6615b5198d8d5c2f6188cd1ef650R88) returns `GRpcOk (OutwardResponseBody "")` instead of a `GRpcOk (Proto.OutwardResponseError (OutwardError RemoteNotFound/DiscoveryFailed _))`. Down the line this seems to call `throwClientError` with a `Servant.DecodeFailure` resulting in 533 with label `federation-invalid-body`

**Comment by smatting** [comment]
https://github.com/wireapp/wire-server/pull/1445/commits/086818f819e84f981c63a8ad56be41fe5bbe8be3 : this shouldn't change the behaviour, but it does. Maybe this is a bug in mu-protobuf or any of its dependencies?

**Comment by jschaul** [comment]
Some of the end2end tests still seem to fail, but strangely only for some endpoints and not all of them. I triggered CI once more just to see if any of it is due to flakyness (though I would assume not)

```
  federation-end2end-user
    lookup user by qualified handle on remote backend:                          OK (0.98s)
    search users on remote backend:               OK (0.47s)
    get users by ids on multiple backends:                                      OK (0.33s)
    claim client prekey:                                                        FAIL
      Exception: Assertions failed
       1: 200 =/= 533
       2: Just (ClientPrekey {prekeyClient = ClientId {client = "1dbfbe22c8a35cb2"}, prekeyData = Prekey {prekeyId = PrekeyId {keyId = 1}, prekeyKey = "pQABAQECoQBYIOjl7hw0D8YRNqkkBQETCxyr7/ywE/2R5RWcUPM+GJACA6EAoQBYILLf1TIwSB62q69Ojs/X1tzJ+dYHNAw4QbW/7TC5vSZqBPY="}}) =/= Nothing

      Response was:
      Response {responseStatus = Status {statusCode = 533, statusMessage = "Unexpected Federation Response"}, responseVersion = HTTP/1.1, responseHeaders = [("Transfer-Encoding","chunked"),("Date","Wed, 21 Apr 2021 17:10:33 GMT"),("Server","Warp/3.3.13"),("Content-Type","application/json")], responseBody = Just "{\"code\":533,\"message\":\"{\\\"code\\\":533,\\\"message\\\":\\\"Could not parse remote federator response: not enough input\\\",\\\"label\\\":\\\"federation-invalid-body\\\"}\",\"label\":\"server-error\"}", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}

      CallStack (from HasCallStack):
        error, called at src/Bilge/Assert.hs:88:5 in bilge-0.22.0-13AJtjUXFk71d9qfIIpaG0:Bilge.Assert
        <!!, called at src/Bilge/Assert.hs:106:19 in bilge-0.22.0-13AJtjUXFk71d9qfIIpaG0:Bilge.Assert
        !!!, called at test/integration/Federation/End2end.hs:137:3 in main:Federation.End2end
    claim prekey bundle:                                                        FAIL
      Exception: Error executing request: HttpExceptionRequest Request {
        host                 = "brig.test-fptbpjobf500.svc.cluster.local"
        port                 = 8080
        secure               = False
        requestHeaders       = [("Z-User","63574a57-2ec6-4ee8-9650-ab815b56d8be")]
        path                 = "users/federation-test-helper.test-fptbpjobf500-fed2.svc.cluster.local/bda6c3ea-6352-4bbb-89c0-6159abb22784/prekeys"
        queryString          = ""
        method               = "GET"
        proxy                = Nothing
        rawBody              = False
        redirectCount        = 10
        responseTimeout      = ResponseTimeoutDefault
        requestVersion       = HTTP/1.1
      }
       (StatusCodeException (Response {responseStatus = Status {statusCode = 533, statusMessage = "Unexpected Federation Response"}, responseVersion = HTTP/1.1, responseHeaders = [("Transfer-Encoding","chunked"),("Date","Wed, 21 Apr 2021 17:10:35 GMT"),("Server","Warp/3.3.13"),("Content-Type","application/json")], responseBody = (), responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}) "{\"code\":533,\"message\":\"{\\\"code\\\":533,\\\"message\\\":\\\"Could not parse remote federator response: not enough input\\\",\\\"label\\\":\\\"federation-invalid-body\\\"}\",\"label\":\"server-error\"}")

      CallStack (from HasCallStack):
        error, called at src/Bilge/Assert.hs:98:18 in bilge-0.22.0-13AJtjUXFk71d9qfIIpaG0:Bilge.Assert
        <!!, called t src/Bilge/Assert.hs:106:19 in bilge-0.22.0-13AJtjUXFk71d9qfIIpaG0:Bilge.Assert
        !!!, called at test/integration/Federation/End2end.hs:158:3 in main:Federation.End2end
    claim multi-prekey bundle:                                                  OK (3.75s)

2 out of 295 tests failed (528.28s)
```

**Review by jschaul — COMMENTED** [review]
About the current end2end test failures: In libs/wire-api-federation/src/Wire/API/Federation/API/Brig.hs  two of the APIs (claimPrekey and getPrekeybundle) are defined with a GET and query parameters. Those need to be changed to an API using POST requests since the POST method is now hardcoded in the runClient function.

File: services/federator/src/Federator/ExternalServer.hs (hunk 2 above)
Comment (jschaul) [reply services/federator/src/Federator/ExternalServer.hs:60]: In a case of diverging versions and an endpoint not yet existing you also need to expect a 404. In general making real http requests and always assuming a 200 isn't realistic.

**Comment by jschaul** [comment]
> Integration tests with unknown domain fail because `gRpcCall` [here](https://github.com/wireapp/wire-server/pull/1445/files#diff-72304808ff3781aece246f66abc167cd8b9b6615b5198d8d5c2f6188cd1ef650R88) returns `GRpcOk (OutwardResponseBody "")` instead of a `GRpcOk (OutwardResponseError (OutwardError RemoteNotFound/DiscoveryFailed _))`. Down the line this seems to call `throwClientError` with a `Servant.DecodeFailure` resulting in 533 with label `federation-invalid-body`

Which tests exactly are affected by this? It might be worth reporting to mu-haskell. (see also https://github.com/higherkindness/mu-haskell/issues/282 for a similar bug)

**Comment by akshaymankar** [comment]
For posterity, the fix for our problem is here: https://github.com/higherkindness/mu-haskell/pull/297
Until then, we can use the workaround.

**Review by jschaul — APPROVED** [review]
Looks good, the remaining TODOs left in the code can be dealt with in subsequent PRs.

**Pull request merged**